br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -642,11 +656,134 @@ func (em *ErrorManager) ReplaceConflictKeys(
+		if err := indexKvRows.Err(); err != nil {
 			return errors.Trace(err)
 		}
 
-		// TODO: check data KV
+		// check data KV
+		dataKvRows, err := em.db.QueryContext(
+			gCtx, fmt.Sprintf(selectDataConflictKeysReplace, em.schemaEscaped),
+			tableName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer dataKvRows.Close()
+		for dataKvRows.Next() {
+			var rawKey []byte
+			if err := dataKvRows.Scan(&rawKey); err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got group raw_key from table",
+				logutil.Key("raw_key", rawKey))
+
+			var mustKeepKvPairs *kv.Pairs
+
+			// get the latest value of rawKey from downstream TiDB
+			latestValue, err := fnGetLatest(gCtx, rawKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if latestValue != nil {
+				handle, err := tablecodec.DecodeRowKey(rawKey)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+					tbl.Meta(), handle, tbl.Cols(), latestValue)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// find out all the KV pairs that are contained in the data KV
+				mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
+			}
+
+			dataKvRowsByRawKey, err := em.db.QueryContext(
+				gCtx, fmt.Sprintf(selectDataConflictByKeysReplace, em.schemaEscaped),
+				tableName, rawKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			for dataKvRowsByRawKey.Next() {
+				var rawValue, rawHandle []byte
+				if err := dataKvRowsByRawKey.Scan(&rawValue, &rawHandle); err != nil {
+					return errors.Trace(err)
+				}
+				em.logger.Debug("got raw_value, raw_handle group by raw_key",
+					zap.Binary("raw_value", rawValue),
+					zap.Binary("raw_handle", rawHandle))
+
+				// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
+				// if not, that means this data KV has been deleted due to overwritten index KV
+				if bytes.Equal(rawValue, latestValue) {
+					continue
+				}
+
+				handle, err := tablecodec.DecodeRowKey(rawHandle)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+					tbl.Meta(), handle, tbl.Cols(), rawValue)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
+				if err != nil {
+					return errors.Trace(err)
+				}
+
+				// find out all the KV pairs that are contained in the data KV
+				kvPairs := encoder.SessionCtx.TakeKvPairs()
+				for _, kvPair := range kvPairs.Pairs {
+					em.logger.Debug("got encoded KV",
+						logutil.Key("key", kvPair.Key),
+						zap.Binary("value", kvPair.Val))
+					kvLatestValue, err := fnGetLatest(gCtx, kvPair.Key)
+					if tikverr.IsErrNotFound(err) || kvLatestValue == nil {
```

Comment:
should we check `IsErrNotFound` for other usage of `fnGetLatest`?

Reply:
I don't think so. Other usage of `fnGetLatest` will not get the result that the key does not exist, if so, we directly return  error.